RDoc-1601: Added folder Client API > Rest API (#1078)
ravendb/docs: Documentation/4.2/Raven.Documentation.Pages/client-api/rest-api/document-commands/batch-commands.markdown

The query string takes [batch options](../../../client-api/rest-api/document-commands/batch-commands#batch-options), which can make the server wait for 
indexing and replication to finish before responding.  

---

### Put Document Command

Upload a new document or update an existing document.  

Format within the `Commands` array in the [request body](../../../client-api/rest-api/document-commands/batch-commands#request-format):  

{CODE-BLOCK: JavaScript}
{
    "Id": "<document ID>",
    "ChangeVector": "<expected change vector>",
    "Document": {
        <document content>
    },
    "Type": "PUT",
    "ForceRevisionCreationStrategy": "<none / before>"
}
{CODE-BLOCK/}

| Parameter | Description | Required |
| - | - | - |
| **Id** | ID of document to create or update | Required |
| **ChangeVector** | If updating an existing document - that document's expected change vector. If it does not match the server-side change vector a concurrency exception is thrown. | Optional |
| **Document** | JSON Document to create, or to replace existing document | Required |
| **Type** | Set to `PUT` | Required |

> | **ForceRevisionCreationStrategy** | Can be set to either `none` or `before` | Optional |

Author: Fixed

| **Id** | Document ID | Required |
| **Name** | Name of attachment to create or update | Required |
| **ContentType** | Mime type of the attachment |
| **ChangeVector** | The document's expected change vector. If it does not match the server-side change vector a concurrency exception is thrown. | Optional |
| **Type** | Set to `AttachmentPUT` | Required |

---

### Delete Attachment Command

Delete an attachment in a certain document.  

Format within the `Commands` array in the [request body](../../../client-api/rest-api/document-commands/batch-commands#request-format):  

{CODE-BLOCK: JavaScript}
{
    "Id": "<document ID>",
    "Name": "<attachment name>",
    "ChangeVector": "<expected change vector>",
    "Type": "AttachmentDELETE"
}
{CODE-BLOCK/}

| Parameter | Description | Required |
| - | - | - |
| **Id** | ID of document for which to delete the attachment | Required |
| Name | Name of the attachment to delete | Required |
| **ChangeVector** | The document's expected change vector. If it does not match the server-side change vector a concurrency exception is thrown. | Optional |
| **Type** | Set to `AttachmentDELETE` | Required |

---

> ### Batch Options

Author: Sounds good. Done

| **numberOfReplicasToWaitFor** | `int` | The number of replicas that should be made before `waitForReplicasTimeout`. Set this parameter to `majority` to wait until the data has been replicated to a majority of the nodes in the database group. Default = `1`. |

{PANEL/}

{PANEL: Response Format}

### Http Status Codes

| Code | Description |
| - | - |
| `201` | The transaction was successfully completed. |
| `408` | The time specified by the options `waitForIndexThrow` or `waitForReplicasTimeout` passed before indexing or replication completed respectively, and an exception was thrown. This only happens if `throwOnTimeoutInWaitForReplicas` or `waitForIndexThrow` are set to `true`. |
| `409` | The specified change vector did not match the server-side change vector, or a change vector was specified for a document that does not exist. A concurrency exception is thrown. |
| `500` | Invalid request, such as a put attachment command for a document that does not exist. |

### Response Body

{CODE-BLOCK: JavaScript}
{
    "Results":[
        { },
        ...
    ]
}
{CODE-BLOCK/}

> * Format within the `Results` array in the response body:  

Reviewer: We need to state here that the returned array of responses is in the order of the received commands

Author: Done

  * [Put Document Command](../../../client-api/rest-api/document-commands/batch-commands#put-document-command-1)  
  * [Patch Document Command](../../../client-api/rest-api/document-commands/batch-commands#patch-document-command-1)  
  * [Delete Document Command](../../../client-api/rest-api/document-commands/batch-commands#delete-document-command-1)  
  * [Delete by Prefix Command](../../../client-api/rest-api/document-commands/batch-commands#delete-by-prefix-command-1)  
  * [Put Attachment Command](../../../client-api/rest-api/document-commands/batch-commands#put-attachment-command-1)  
  * [Delete Attachment Command](../../../client-api/rest-api/document-commands/batch-commands#delete-attachment-command-1)  

---

### Put Document Command

{CODE-BLOCK: JavaScript}
{
    "Type": "PUT",
    "@id": "<document ID>",
    "@collection": "<collection name>",
    "@change-vector": "<current change vector>",
    "@last-modified": "<date and time UTC>"
}
{CODE-BLOCK/}

| Parameter | Description |
| - | - | - |
| **Type** | Same as the `Type` of the command sent - in this case `PUT`. |

> | **@id** | The ID of the document that has been created or deleted. |

Author: Fixed

| **@collection** | Name of the [collection](../../../client-api/faq/what-is-a-collection) that contains the document. If none was specified, the collection will be `@empty`. |
| **@change-vector** | The document's change vector after the command was executed. |

---

### Patch Document Command

{CODE-BLOCK: JavaScript}
{
    "@id": "<document ID>",
    "@change-vector": "<current change vector>",
    "@last-modified": "<date and time UTC>",
    "Type": "PATCH",
    "PatchStatus": "<patch status>",
    "Debug": null
}
{CODE-BLOCK/}

| Parameter | Description |
| - | - | - |

> | **@id** | The ID of the document that has been created or deleted. |

Reviewer: ```suggestion
| **@id** | The ID of the document that has been created. |
```

| **Id** | The ID of the document that has been created or deleted. |
| **Type** | Same as the `Type` of the command sent - in this case `DELETE`. |
| **Deleted** | `true` if the document was successfully deleted, `false` if not (for instance, because the specified document did not exist). |

---

### Delete by Prefix Command

{CODE-BLOCK: JavaScript}
{
    "Id": "<prefix>",
    "Type": "DELETE",
    "Deleted": <boolean>
}
{CODE-BLOCK/}

| Parameter | Description |
| - | - | - |
| **Id** | The document ID prefix of the documents which were deleted. |
| **Type** | Same as the `Type` of the command sent - in this case `DELETE`. |
| **Deleted** | `true` if the documents were successfully deleted, `false` if not (for instance, because no documents with the specified prefix exist). |

---

### Put Attachment Command

{CODE-BLOCK: JavaScript}
{
    "Id": "<document ID>",
    "Type": "AttachmentPUT",
    "Name": "<attachment name>",
    "ChangeVector": "<attachment change vector>",
    "Hash": "<Hash>",
    "ContentType": "<MIME type>",
    "Size": <attachment size in bytes>,
    "DocumentChangeVector": "<current change vector>"
}
{CODE-BLOCK/}

| Parameter | Description |
| - | - | - |

> | **Id** | The document ID prefix of the documents which were deleted. |

Reviewer: ```suggestion
| **Id** | The document ID of the document to which the attachment belongs. |
```

| **Type** | Same as the `Type` of the command sent - in this case `AttachmentPUT`. |
| **Name** | Name of the attachment that was created or updated. |
| **ChangeVector** | A change vector specific to the _attachment_, distinct from the usual document change vector. Use this change vector in requests to update this attachment. |
| **Hash** | Hash representing the attachment. |
| **ContentType** | MIME type of the attachment. |
| **Size** | Size of the attachment in bytes. |
| **DocumentChangeVector** | The document's change vector after the command was executed. |

---

### Delete Attachment Command

{CODE-BLOCK: JavaScript}
{
    "Type": "AttachmentDELETE",
    "@id": "<document ID>",
    "Name": "<attachment name",
    "DocumentChangeVector": "<current change vector>"
}
{CODE-BLOCK/}

| Parameter | Description |
| - | - | - |
| **Type** | Same as the `Type` of the command sent - in this case `AttachmentDELETE`. |

> | **@id** | The ID of the documents that was deleted. |

Reviewer: ```suggestion
| **@id** | The ID of the documents to which the deleted attachment belongs. |
```
